### PR TITLE
enhance(telemetry): measure all external link clicks

### PR DESCRIPTION
**Pull request by LeoMcA**
## Summary

https://mozilla-hub.atlassian.net/browse/MP-1631

### Problem

- two issues which meant we got an incomplete picture of our external links:
  - `glean-data` would mean the `external-link` ping wouldn't get send
  - external links without `class="external"` wouldn't have a ping sent

### Solution

- no `else if`
- check `href` rather than `class`

---

## How did you test this change?

```
REACT_APP_GLEAN_ENABLED=true
REACT_APP_GLEAN_DEBUG=true
```
in `.env`
`yarn dev`
localhost:3000
clicked various internal and external links, ensured:
- external links without `class="external"` got a ping sent
- internal links got no ping sent

### CODE REVIEW

File: client/src/telemetry/glean-context.tsx [hunk 1]
```diff
@@ -174,8 +174,12 @@ function handleLinkClick(ev: MouseEvent, click: (source: string) => void) {
   if (anchor instanceof HTMLAnchorElement) {
     if (anchor.dataset.glean) {
       click(anchor.dataset.glean);
-    } else if (anchor.classList.contains("external")) {
-      click(`external-link: ${anchor.getAttribute("href") || ""}`);
+    }
+    if (anchor.href) {
+      const url = new URL(anchor.href, document.location.href);
```
Comment (caugner) [inline client/src/telemetry/glean-context.tsx:179]: What are the chances of `anchor.href` containing an invalid URL, triggering an error?

**Review by caugner — CHANGES_REQUESTED** [review]

File: client/src/telemetry/glean-context.tsx (hunk 1 above)
Comment (fiji-flo) [reply client/src/telemetry/glean-context.tsx:179]: Can't wait for `URL.parse` to be baseline green. But yes let's add a `try...catch`

File: client/src/telemetry/glean-context.tsx [hunk 2]
```diff
@@ -174,8 +174,12 @@ function handleLinkClick(ev: MouseEvent, click: (source: string) => void) {
...
+      if (url.origin !== document.location.origin) {
+        click(`external-link: ${url.href}`);
```
[the remark this answers is not in the document]
Comment (LeoMcA) [reply client/src/telemetry/glean-context.tsx:181]: That's exactly why I used it: we want normalized url values in glean, not raw values, e.g. `https://example.com` `https://example.com/` and `//example.com` shouldn't be treated as different external links: they're the same, or e.g. `https://example.com/foo bar` and `https://example.com/foo%20bar    `: using `url.href` normalizes them all to `https://example.com` and `https://example.com/foo%20bar` respectively

[the remark this answers is not in the document]
Comment (caugner) [reply client/src/telemetry/glean-context.tsx:181]: Okay, that seems reasonable. We actually had slight variations of the Scrimba URLs on the Curriculum landing page, which would have been fixed by this. So 👍

[the remark this answers is not in the document]
Comment (caugner) [reply client/src/telemetry/glean-context.tsx:187]: FWIW It looks like the href gets parsed in Firefox, so `anchor.href` doesn't have the "raw" value:

https://developer.mozilla.org/en-US/play?id=F%2BV7lqnurjSRhYLrIydkUs2eK55MnyrSmjNE%2FvJXIv3j3zM0pmTzqeBrBKqy%2Fu9SFzMWXSb7UvDq4h3D

[the remark this answers is not in the document]
Comment (caugner) [reply client/src/telemetry/glean-context.tsx:187]: https://developer.mozilla.org/en-US/play?id=DYidtfiBwXFRx9RL449JNvG4cnHtq0ec1QQBeHgBCmkpm5h0jPXIeMiWjNLZPWrJ47yTUDlqqja2U9gG

[the remark this answers is not in the document]
Comment (caugner) [reply client/src/telemetry/glean-context.tsx:187]: I guess it means it doesn't matter.

File: client/src/telemetry/glean-context.tsx [hunk 3]
```diff
@@ -174,8 +174,18 @@ function handleLinkClick(ev: MouseEvent, click: (source: string) => void) {
...
+    if (anchor.href) {
+      let url: URL | undefined;
+      try {
+        url = new URL(anchor.href, document.location.href);
+      } catch {
+        click(`invalid-link: ${anchor.href}`);
+      }
+      if (url && url.origin !== document.location.origin) {
+        // use normalized href from `url`:
+        click(`external-link: ${url.href}`);
```
Comment (caugner) [inline client/src/telemetry/glean-context.tsx:187]: While we're here, could we extract those constants to the [`telemetry/constants.ts`](https://github.com/mdn/yari/blob/main/client/src/telemetry/constants.ts) file?

**Review by caugner — CHANGES_REQUESTED** [review]

**Review by caugner — APPROVED** [review]
LGTM, just [one nit](https://github.com/mdn/yari/pull/11981/files#r1806460853).

**Review by caugner — APPROVED** [review]

**Pull request merged**